sys/behave/src/Unfold.hs

```diff
@@ -55,88 +55,81 @@ import qualified Eval
 unfoldCTbranch chsets (CTpref ctoffs cthidvars' ctpreds' ctnext')
-   | Set.null ctoffs && null cthidvars'  =  do                -- tau action or nothing
-       predVal <- Eval.evalCnrs ctpreds'
-       if  predVal
-         then do nextcnode <- return $ nextNode Map.empty ctnext'
-                 nextctree <- expand chsets nextcnode
-                 nextbtree <- unfoldCT chsets nextctree
-                 return $ [ BTtau nextbtree ]
-         else do return $ []
-   | Set.null ctoffs && not (null cthidvars') =              -- tau action or nothing
-       let assertions = foldr Solve.add Solve.empty ctpreds'
-        in do
-           smtEnv <- IOB.getSMT "current"
-           (sp,smtEnv') <- lift $ runStateT (Solve.uniSolve cthidvars' assertions) smtEnv
-           IOB.putSMT "current" smtEnv'
-           case sp of
-           { SolveDefs.Solved sol    -> do nextcnode <- return $ nextNode sol ctnext'
-                                           nextctree <- expand chsets nextcnode
-                                           nextbtree <- unfoldCT chsets nextctree
-                                           return $ [ BTtau nextbtree ]
-           ; SolveDefs.Unsolvable    -> do return $ []
-           ; SolveDefs.UnableToSolve -> do IOB.putMsgs [ EnvData.TXS_CORE_USER_ERROR
-                                                         $ "unfoldCTbranch: Not unique" ]
-                                           return $ [ BTpref ctoffs cthidvars' ctpreds' ctnext' ]
-           }
-   | True =  -- not $ Set.null ctoffs                                  -- visible action or nothing
+   | Set.null ctoffs =                                           -- tau action or nothing
+        if null cthidvars'
+          then do
+            predVal <- Eval.evalCnrs ctpreds'
+            if  predVal
+              then let nextcnode = nextNode Map.empty ctnext'
+                     in do nextctree <- expand chsets nextcnode
+                           nextbtree <- unfoldCT chsets nextctree
+                           return [ BTtau nextbtree ]
+              else return []
+          else
+            let assertions = foldr Solve.add Solve.empty ctpreds'
+              in do
+                smtEnv <- IOB.getSMT "current"
+                (sp,smtEnv') <- lift $ runStateT (Solve.uniSolve cthidvars' assertions) smtEnv
+                IOB.putSMT "current" smtEnv'
+                case sp of
+                  SolveDefs.Solved sol    -> do let nextcnode = nextNode sol ctnext'
+                                                nextctree <- expand chsets nextcnode
+                                                nextbtree <- unfoldCT chsets nextctree
+                                                return [ BTtau nextbtree ]
+                  SolveDefs.Unsolvable    -> return []
+                  SolveDefs.UnableToSolve -> do IOB.putMsgs [ EnvData.TXS_CORE_USER_ERROR "unfoldCTbranch: Not unique" ]
```

Comment:
I personally prefer to put the `do` actions in a line below, like this:

```haskell
  SolveDefs.UnableToSolve -> 
    do IOB.putMsgs [ EnvData.TXS_CORE_USER_ERROR "unfoldCTbranch: Not unique" ]
```
because in this way we can use less scree space.